model/riscv_insts_zaamo.sail

```diff
@@ -34,33 +44,37 @@ mapping encdec_amoop : amoop <-> bits(5) = {
   AMOMIN  <-> 0b10000,
   AMOMAX  <-> 0b10100,
   AMOMINU <-> 0b11000,
-  AMOMAXU <-> 0b11100
+  AMOMAXU <-> 0b11100,
+  AMOCAS  <-> 0b00101,
 }
 
 mapping clause encdec = AMO(op, aq, rl, rs2, rs1, size, rd)
-  <-> encdec_amoop(op) @ bool_bits(aq) @ bool_bits(rl) @ encdec_reg(rs2) @ encdec_reg(rs1) @ 0b0 @ size_enc(size) @ encdec_reg(rd) @ 0b0101111
-  when currentlyEnabled(Ext_Zaamo) & amo_width_valid(size)
+  <-> encdec_amoop(op) @ bool_bits(aq) @ bool_bits(rl) @ encdec_reg(rs2) @ encdec_reg(rs1) @ size_enc_wide(size) @ encdec_reg(rd) @ 0b0101111
+  when amo_encoding_valid(size, op, rs2, rd)
 
 /* NOTE: Currently, we only EA if address translation is successful.
    This may need revisiting. */
 function clause execute (AMO(op, aq, rl, rs2, rs1, width, rd)) = {
   let 'width = width;
 
   // This is checked during decoding.
-  assert(width <= xlen_bytes);
+  assert(width <= xlen_bytes * 2);
 
   /* Get the address, X(rs1) (no offset).
     * Some extensions perform additional checks on address validity.
     */
   match ext_data_get_addr(rs1, zeros(), ReadWrite(Data, Data), width) {
     Ext_DataAddr_Error(e)  => Ext_DataAddr_Check_Failure(e),
     Ext_DataAddr_OK(vaddr) => {
-      if not(is_aligned_addr(bits_of(vaddr), width))
+      if not(is_aligned_addr(vaddr, width))
       then Memory_Exception(vaddr, E_SAMO_Addr_Align())
       else match translateAddr(vaddr, ReadWrite(Data, Data)) {
         Err(e, _) => Memory_Exception(vaddr, e),
         Ok(addr, _) => {
-          let rs2_val = X(rs2)[width * 8 - 1 .. 0];
+          // AMOCAS.D for RV32 and AMOCAS.Q for RV64 operate on a pair of
+          // registers. When the first register of a register pair is
+          // x0, then both halves of the pair read/write as zero.
+          let rs2_val = X_pair(rs2, width)[width * 8 - 1 .. 0];
```

Comment:
Wouldn't this be readable as a call to truncate?
```suggestion
          let rs2_val = truncate(if width > xlen_bytes then X_pair(rs2) else X(rs2), width * 8);
```

Reply:
Ah yeah these suggestions seem reasonable to me. I'll update the code.